### DISCUSSION TITLE
Convert SerialReceived Hex to Int on ESP8266

### DISCUSSION THREAD

**Discussion by patrmis** [arendst/Tasmota#21753] [Q&A]
I would like to connect an SR04 ultrasonic Sensor via serial on an ESP8266.

I have my rule1 set to `ON Time#minute DO serialsend5 0xA0 ENDON"` to which I receive a response such as `{"SerialReceived":"16059C"}`.

Of course I could post process the hex on the receiving end of the MQTT, but I'd prefer to do it on the device and transmit the Int instead. All the posts I come across for Hex to Dec/Int conversion rely on Berry, but it's my understanding that the ESP8266 is not Berry capable.

Can this be achieved via a rule or other means on ESP8266?

I realise I could use the Trig/Echo method, but I find I get better stability with the serial comms.
Plus a future project will require 2 SR04, at which stage I'm hoping to connect one via serial and the other via trig/echo.

Specifications:
Wemos D1 Mini
Tasmota 14.1.0
RCWL-1655

**Comment by sfromis** [comment]
On ESP8266, "easy" options are limited. Rules has no support for hex, but maybe you could do something with the scripter, which does seem to support such. Of course, you could also make a native driver.

Anyway, the trig/echo method has been extended to support up to 3 SR04.

What I'd prefer would be to do that kind of thing on ESP32, where the Berry language has good support for such toys, better serial communication than the serial bridge (also with more than one hardware serial), and also common programming features like handling hex data.
